File: baekjoon_all/01000+/boj_1182.py
# Solve 2023-03-29

import sys

input = lambda : sys.stdin.readline().rstrip()

def dfs(depth = 0, val = 0, cnt = 0):
    if depth == n:
        if val == s and cnt > 0:
            return 1
        return 0

    res = dfs(depth + 1, val, cnt)
    res += dfs(depth + 1, val + li[depth], cnt + 1)
    return res

n, s = map(int, input().split())
li = list(map(int, input().split()))
ans = dfs()

print(ans)

File: baekjoon_all/01000+/boj_1182.java
// Solve 2024-02-03

// 백준에 제출할 때는 class 이름을 Main으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class boj_1182 {

    static int n, s, ans;
    static int[] arr;

    static void recur(int depth, int sum, int cnt) {
        if (depth == n) {
            if (sum == s && cnt > 0) {
                ans++;
            }

            return;
        }

        recur(depth + 1, sum, cnt);
        recur(depth + 1, sum + arr[depth], cnt + 1);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        s = Integer.parseInt(st.nextToken());
        arr = new int[n];
        st = new StringTokenizer(br.readLine(), " ");

        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        recur(0, 0, 0);

        System.out.println(ans);
        br.close();
    }

}
